pkg/api/annotations.go

```diff
@@ -478,9 +444,23 @@ func AnnotationTypeScopeResolver() (string, accesscontrol.AttributeScopeResolveF
 	return accesscontrol.ScopeAnnotationsProvider.GetResourceScope(""), annotationTypeResolver
 }
 
-func (hs *HTTPServer) canCreateOrganizationAnnotation(c *models.ReqContext) (bool, error) {
-	evaluator := accesscontrol.EvalPermission(accesscontrol.ActionAnnotationsCreate, accesscontrol.ScopeAnnotationsTypeOrganization)
-	return hs.AccessControl.Evaluate(c.Req.Context(), c.SignedInUser, evaluator)
+func (hs *HTTPServer) canCreateAnnotation(c *models.ReqContext, dashboardId int64) (bool, error) {
+	if dashboardId != 0 {
+		if hs.Features.IsEnabled(featuremgmt.FlagAccesscontrol) {
+			evaluator := accesscontrol.EvalPermission(accesscontrol.ActionAnnotationsCreate, accesscontrol.ScopeAnnotationsTypeDashboard)
+			if canSave, err := hs.AccessControl.Evaluate(c.Req.Context(), c.SignedInUser, evaluator); err != nil || !canSave {
+				return canSave, err
+			}
+		}
+		return canEditDashboard(c, dashboardId)
```

Comment:
great job on the PR. I feel there could be some comment to describe why inside createAnnotation we have `canEditDashboard`.

> checks that user has permissions to create dashboard annotations if RBAC is enabled (previously users who only had permissions to create organization annotations could also create dashboard annotations);

I am reading this, but to me it does not explain why you need save, or edit.

hope this makes sense

Reply:
This is the way things were set up with legacy access control as well - to add an annotation to a dashboard, you need edit access to this dashboard. I feel like it kind of makes sense, as adding an annotation changes the way the dashboard looks.

Do you think it'd be helpful to add this information in a comment?

Reply:
i thought i would take a shot at refactoring, as i did not understand what was going on.

But now i see it, it is still a bit messy to me but workable.

wdyt?
```go
	// ---
	// scoped dashboard annotations
	// --

	// accesscontrol
	if hs.Features.IsEnabled(featuremgmt.FlagAccesscontrol) && dashboardId != 0 {
		evaluator := accesscontrol.EvalPermission(accesscontrol.ActionAnnotationsCreate, accesscontrol.ScopeAnnotationsTypeDashboard)
		if canSave, err := hs.AccessControl.Evaluate(c.Req.Context(), c.SignedInUser, evaluator); err != nil || !canSave {
			return canSave, err
		}
	}

	// no accesscontrol
	if !hs.Features.IsEnabled(featuremgmt.FlagAccesscontrol) && dashboardId != 0 {
		return canEditDashboard(c, dashboardId)
	}

	// --
	// organization annotations
	//

	// accesscontrol
	if hs.Features.IsEnabled(featuremgmt.FlagAccesscontrol) {
		return c.SignedInUser.HasRole(models.ROLE_EDITOR), nil
	}

	return c.SignedInUser.HasRole(models.ROLE_EDITOR), nil
```